src/io/der.rs

```diff
@@ -121,8 +121,15 @@ where
     inner.read_all(error, decoder)
 }
 
-fn nonnegative_integer<'a>(
+fn read_nonnegative_integer<'a>(
     input: &mut untrusted::Reader<'a>, min_value: u8,
+) -> Result<untrusted::Input<'a>, error::Unspecified> {
+    let value = expect_tag_and_get_value(input, Tag::Integer)?;
+    nonnegative_integer(value, min_value)
+}
+
+pub fn nonnegative_integer<'a>(
```

Comment:
This wouldn't belong in the `der` submodule any more since it doesn't have anything to do with DER anymore.

One possibility: 1. Move all this logic to the new `Positive` constructor you added. 2. Rewrite `der::nonnegative_integer` to handle two cases separately: the zero-value case and the other case.